File: SiteBase/Scripts/SessionAuditor.cs
// ---------------------------------------------------------------------- //
//                                                                        //
//                       Copyright (c) 2007-2014                          //
//                         Digital Beacon, LLC                            //
//                                                                        //
// ---------------------------------------------------------------------- //

using System;
using System.Collections.Generic;
using System.Html;
using System.Collections;
using jQueryLib;


namespace DigitalBeacon.SiteBase
{
	/// <summary>
	/// Polls session status and redirects to login page when session expires
	/// </summary>
	public class SessionAuditor
	{
		public const int DefaultPollingInterval = 300;
		public const int DefaultNotificationTime = 60;

		private const string DefaultAuthSessionUrl = "~/.authSession.ashx";
		private const string DefaultExtendSessionUrl = "~/account/extendSession";
		private const string DefaultSignOutUrl = "~/identity/signOut";
		private const string DefaultNotificationHeading = "Session Expiration";
		private const string DefaultNotificationMessage = "Your session is set to expire in one minute.";
		private const string DefaultExtendButtonText = "Extend Session";

		private static SessionAuditor _instance;

		private string _authSessionUrl;
		private string _extendSessionUrl;
		private string _signOutUrl;
		private int _intervalTimerId;
		private int _notificationTimerId;
		private int _signOutTimerId;
		private int _pollingInterval;
		private int _notificationTime;
		private bool _displayNotification = true;
		private string _notificationHeading;
		private string _notificationMessage;
		private string _extendButtonText;

		public static SessionAuditor instance()
		{
			if (_instance == null)
			{
				_instance = new SessionAuditor();
			}
			return _instance;
		}

		private SessionAuditor()
		{
		}

		public int pollingInterval
		{
			get { return _pollingInterval > 0 ? _pollingInterval : DefaultPollingInterval; }
			set { _pollingInterval = value; }
		}

		public int notificationTime
		{
			get { return _notificationTime > 0 ? _notificationTime : DefaultNotificationTime; }
			set { _notificationTime = value; }
		}

		public bool displayNotification
		{
			get { return _displayNotification; }
			set { _displayNotification = value; }
		}

		public string authSessionUrl
		{
			get { return digitalbeacon.resolveUrl((dynamic)_authSessionUrl || DefaultAuthSessionUrl); }
			set { _authSessionUrl = value; }
		}

		public string extendSessionUrl
		{
			get { return digitalbeacon.resolveUrl((dynamic)_extendSessionUrl || DefaultExtendSessionUrl); }
			set { _extendSessionUrl = value; }
		}

		public string signOutUrl
		{
			get { return digitalbeacon.resolveUrl((dynamic)_signOutUrl || DefaultSignOutUrl); }
			set { _signOutUrl = value; }
		}

		public string notificationHeading
		{
			get { return (dynamic)_notificationHeading || DefaultNotificationHeading; }
			set { _notificationHeading = value; }
		}

		public string notificationMessage
		{
			get { return (dynamic)_notificationMessage || DefaultNotificationMessage; }
			set { _notificationMessage = value; }
		}

		public string extendButtonText
		{
			get { return (dynamic)_extendButtonText || DefaultExtendButtonText; }
			set { _extendButtonText = value; }
		}

		public void start()
		{
			stop();
			_intervalTimerId = window.setInterval((Action)(() => { checkSession(); }), pollingInterval * 1000);
		}

		public void stop()
		{
			if (_intervalTimerId > 0)
			{
				window.clearInterval(_intervalTimerId);
				_intervalTimerId = 0;
			}
			if (_notificationTimerId > 0)
			{
				window.clearTimeout(_notificationTimerId);
				_notificationTimerId = 0;
			}
			if (_signOutTimerId > 0)
			{
				window.clearTimeout(_signOutTimerId);
				_signOutTimerId = 0;
			}
		}

		private void checkSession()
		{
			jQuery.post(authSessionUrl, null, (data, status, xhr) => { checkSessionOnSuccess(data); });
		}

		private void checkSessionOnSuccess(object data)
		{
			var secondsRemaining = (int)data;
			digitalbeacon.log(secondsRemaining + " seconds remaining...");
			if (_displayNotification && _notificationTimerId <= 0 && secondsRemaining < notificationTime + pollingInterval)
			{
				_notificationTimerId = window.setTimeout((Action)(() => { showNotification(); }), (secondsRemaining - notificationTime) * 1000);
				_signOutTimerId = window.setTimeout((Action)(() => { signOut(); }), secondsRemaining * 1000);
				window.clearInterval(_intervalTimerId);
				_intervalTimerId = 0;
			}
		}

		private void showNotification()
		{
			jQuery.post(authSessionUrl, null, (data, status, xhr) =>
			{
				var secondsRemaining = (int)data;
				if (secondsRemaining < notificationTime + pollingInterval)
				{
					sitebase.displayMessage(
						notificationMessage.formatWith(notificationTime, notificationTime / 60),
						notificationHeading, "", extendButtonText, () => { extendSession(); });
				}
				else
				{
					// session was extended
					start();
				}
			});
		}

		private void signOut()
		{
			jQuery.post(authSessionUrl, null, (data, status, xhr) =>
			{
				int secondsRemaining = (int)data;
				if (secondsRemaining < pollingInterval)
				{
					window.location.href = signOutUrl;
				}
				else
				{
					// session was extended
					start();
				}
			});
		}

		private void extendSession()
		{
			jQuery.post(extendSessionUrl);
			start();
		}
	}
}

File: SiteBase/Web/Scripts/SessionAuditor.ts
// ---------------------------------------------------------------------- //
//                                                                        //
//                       Copyright (c) 2007-2014                          //
//                         Digital Beacon, LLC                            //
//                                                                        //
// ---------------------------------------------------------------------- //

namespace DigitalBeacon.SiteBase {
    /// <summary>
    /// Polls session status and redirects to login page when session expires
    /// </summary>
    class SessionAuditor {
        
        public static readonly DefaultPollingInterval: number = 300;
        public static readonly DefaultNotificationTime: number = 60;

        private static readonly DefaultAuthSessionUrl: string = "~/.authSession.ashx";
        private static readonly DefaultExtendSessionUrl: string = "~/account/extendSession";
        private static readonly DefaultSignOutUrl: string = "~/identity/signOut";
        private static readonly DefaultNotificationHeading: string = "Session Expiration";
        private static readonly DefaultNotificationMessage: string = "Your session is set to expire in one minute.";
        private static readonly DefaultExtendButtonText: string = "Extend Session";

        // private static SessionAuditor _instance;

        private _authSessionUrl: string;
        private _extendSessionUrl: string;
        private _signOutUrl: string;
        private _intervalTimerId: number;
        private _notificationTimerId: number;
        private _signOutTimerId: number;
        private _pollingInterval: number;
        private _notificationTime: number;
        private _displayNotification: boolean = true;
        private _notificationHeading: string;
        private _notificationMessage: string;
        private _extendButtonText: string;

        // public static SessionAuditor instance()
        // {
        // 	if (_instance == null)
        // 	{
        // 		_instance = new SessionAuditor();
        // 	}
        // 	return _instance;
        // }
        //
        // private SessionAuditor()
        // {
        // }

        get authSessionUrl(): string {
            return this._authSessionUrl;
        }

        set authSessionUrl(value: string) {
            this._authSessionUrl = value;
        }

        get extendSessionUrl(): string {
            return this._extendSessionUrl;
        }

        set extendSessionUrl(value: string) {
            this._extendSessionUrl = value;
        }

        get signOutUrl(): string {
            return this._signOutUrl;
        }

        set signOutUrl(value: string) {
            this._signOutUrl = value;
        }

        get pollingInterval(): number {
            return this._pollingInterval;
        }

        set pollingInterval(value: number) {
            this._pollingInterval = value;
        }

        get notificationTime(): number {
            return this._notificationTime;
        }

        set notificationTime(value: number) {
            this._notificationTime = value;
        }

        get displayNotification(): boolean {
            return this._displayNotification;
        }

        set displayNotification(value: boolean) {
            this._displayNotification = value;
        }

        get notificationHeading(): string {
            return this._notificationHeading;
        }

        set notificationHeading(value: string) {
            this._notificationHeading = value;
        }

        get notificationMessage(): string {
            return this._notificationMessage;
        }

        set notificationMessage(value: string) {
            this._notificationMessage = value;
        }

        get extendButtonText(): string {
            return this._extendButtonText;
        }

        set extendButtonText(value: string) {
            this._extendButtonText = value;
        }

        get notificationTimerId(): number {
            return this._notificationTimerId;
        }

        set notificationTimerId(value: number) {
            this._notificationTimerId = value;
        }

        get signOutTimerId(): number {
            return this._signOutTimerId;
        }

        set signOutTimerId(value: number) {
            this._signOutTimerId = value;
        }

        get intervalTimerId(): number {
            return this._intervalTimerId;
        }

        set intervalTimerId(value: number) {
            this._intervalTimerId = value;
        }

        public start() {
            stop();
            this._intervalTimerId = setInterval(() => {
                this.checkSession();
            }, this.pollingInterval * 1000);
        }

        public stop() {
            if (this._intervalTimerId > 0) {
                clearInterval(this._intervalTimerId);
                this._intervalTimerId = 0;
            }
            if (this._notificationTimerId > 0) {
                clearTimeout(this._notificationTimerId);
                this._notificationTimerId = 0;
            }
            if (this._signOutTimerId > 0) {
                clearTimeout(this._signOutTimerId);
                this._signOutTimerId = 0;
            }
        }

        checkSession() {
            jQuery.post(this.authSessionUrl, null, (data, status, xhr) => {
                this.checkSessionOnSuccess(data);
            });
        }

        checkSessionOnSuccess(data: number) {
            let secondsRemaining = data;
            digitalbeacon.log(secondsRemaining + " seconds remaining...");
            if (this._displayNotification && this._notificationTimerId <= 0 && secondsRemaining < this.notificationTime + this.pollingInterval) {
                this._notificationTimerId = window.setTimeout(() => {
                    this.showNotification();
                }, (secondsRemaining - this.notificationTime) * 1000);
                this._signOutTimerId = window.setTimeout(() => {
                    this.signOut();
                }, secondsRemaining * 1000);
                clearInterval(this._intervalTimerId);
                this._intervalTimerId = 0;
            }
        }

        showNotification() {
            jQuery.post(this.authSessionUrl, null, (data, status, xhr) => {
                var secondsRemaining = data;
                if (secondsRemaining < this.notificationTime + this.pollingInterval) {
                    sitebase.displayMessage(
                        this.notificationMessage.formatWith(this.notificationTime, this.notificationTime / 60),
                        this.notificationHeading, "", this.extendButtonText, () => {
                            this.extendSession();
                        });
                } else {
                    // session was extended
                    this.start();
                }
            });
        }

        signOut() {
            jQuery.post(this.authSessionUrl, null, (data, status, xhr) => {
                let secondsRemaining = data;
                if (secondsRemaining < this.pollingInterval) {
                    window.location.href = this.signOutUrl;
                } else {
                    // session was extended
                    this.start();
                }
            });
        }

        extendSession() {
            jQuery.post(this.extendSessionUrl);
            this.start();
        }
    }
}
